contrib/endpoints/src/api_manager/config_manager.cc

```diff
@@ -0,0 +1,302 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "contrib/endpoints/src/api_manager/config_manager.h"
+#include "contrib/endpoints/src/api_manager/utils/marshalling.h"
+#include "utils/marshalling.h"
+
+#include "google/api/servicemanagement/v1/servicemanager.pb.h"
+#include "utils/md5.h"
+
+#include <math.h>
+
+#include <chrono>
+#include <ctime>
+#include <future>
+#include <iostream>
+#include <limits>
+#include <mutex>
+#include <regex>
+#include <thread>
+
+using ::google::api::Service;
+using ::google::api::servicemanagement::v1::ListServiceRolloutsResponse;
+using ::google::api::servicemanagement::v1::Rollout_RolloutStatus;
+
+namespace google {
+namespace api_manager {
+
+namespace {
+
+const char kDelimiter[] = "\0";
+const int kDelimiterLength = 1;
+
+// Initial metadata fetch timeout (1s)
+const int kInceptionFetchTimeout = 1000;
+// Maximum number of retries to fetch metadata
+const int kInceptionFetchRetries = 5;
+// Default rollouts refresh interval in ms
+const int kConfigUpdateCheckInterval = 60000;
+
+const char kRolloutStrategyManaged[] = "managed";
+// Default service management API url
+const char kServiceManagementService[] =
+    "https://servicemanagement.googleapis.com";
+const char kServiceManagementServiceManager[] =
+    "/google.api.servicemanagement.v1.ServiceManager";
+// config_id validation
+const char kValidConfigDateFormat[] = "%Y-%m-%d";
+const char kConfigRevisionDelimeter = 'r';
+// static configs for error handling
+static std::vector<std::pair<std::string, int>> empty_configs;
+
+// Check config_id is valid
```

Comment:
I am not sure if this format is fixed.  Inception may change anytime.   Please check with Adam.

My suggestion is we don't need to validate it.     As long as we could fetch it,  it is good.

If we do want to validate, please add comment on the config_id format.